### PR TITLE
expression: support JSON for builtin function COALESCE

### PR DESCRIPTION
**Pull request by zz-jason**
Signed-off-by: Jian Zhang <zjsariel@gmail.com>

### What problem does this PR solve? 

Support JSON type for builtin function COALESCE, to avoid panic once we returned a nil pointer.

### What is changed and how it works?

As the above said.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by eurekaka** [comment]
@zz-jason CI failed, PTAL

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -339,6 +342,28 @@ func (b *builtinCoalesceDurationSig) evalDuration(row chunk.Row) (res types.Dura
 	return res, isNull, err
 }
 
+// builtinCoalesceJsonSig is buitin function coalesce signature which return type time
```
Comment (lamxTyler) [inline expression/builtin_compare.go]: `return type json`?

File: expression/integration_test.go [hunk 2]
```diff
@@ -3857,3 +3857,20 @@ func (s *testIntegrationSuite) TestCastAsTime(c *C) {
 	err = tk.ExecToErr(`select cast(col5 as time(31)) from t where col1 is null;`)
 	c.Assert(err.Error(), Equals, "[expression:1426]Too big precision 31 specified for column 'CAST'. Maximum is 6.")
 }
+
+func (s *testIntegrationSuite) TestCoalesce(c *C) {
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: These tests can be added in line 2840

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -202,6 +202,9 @@ func (c *coalesceFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 		}
 		sig = &builtinCoalesceDurationSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_CoalesceDuration)
+	case types.ETJson:
+		sig = &builtinCoalesceJsonSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CoalesceJson)
```
Comment (eurekaka) [inline expression/builtin_compare.go]: We need to update `getSignatureByPB` as well.

File: expression/builtin_compare.go [hunk 4]
```diff
@@ -339,6 +342,28 @@ func (b *builtinCoalesceDurationSig) evalDuration(row chunk.Row) (res types.Dura
 	return res, isNull, err
 }
 
+// builtinCoalesceJsonSig is buitin function coalesce signature which return type time
+// See http://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_coalesce
+type builtinCoalesceJsonSig struct {
```
Comment (eurekaka) [inline expression/builtin_compare.go]: lint tool suggests "type builtinCoalesceJsonSig should be builtinCoalesceJSONSig"

**Pull request merged**